modules/AlphaWalletOpenSea/AlphaWalletOpenSea/OpenSea.swift

```diff
@@ -19,72 +20,203 @@ public protocol OpenSeaDelegate: AnyObject {
 }
 
 public enum OpenSeaApiError: Error {
+    case `internal`(Error)
+    case invalidJson
     case rateLimited
     case invalidApiKey
     case expiredApiKey
 }
 
-public class OpenSea {
-    public static var isLoggingEnabled = false
+extension OpenSeaApiError {
+    init(error: Error) {
+        if let e = error as? OpenSeaApiError {
+            self = e
+        } else {
+            self = .internal(error)
+        }
+    }
+}
+
+extension URLRequest {
+    public typealias Response = (data: Data, response: HTTPURLResponse)
+}
+
+public typealias Request = Alamofire.URLRequestConvertible
+
+public protocol Networking {
+    func send(request: Request) -> AnyPublisher<URLRequest.Response, PromiseError>
+}
+
+final class OpenSeaRetryPolicy: RetryPolicy {
+
+    init() {
+        super.init(retryableHTTPStatusCodes: Set([429, 408, 500, 502, 503, 504]))
+    }
+    
+    override func retry(_ request: Alamofire.Request,
+                        for session: Session,
+                        dueTo error: Error,
+                        completion: @escaping (RetryResult) -> Void) {
+
+        if request.retryCount < retryLimit, shouldRetry(request: request, dueTo: error) {
+            if let httpResponse = request.response, let delay = OpenSeaRetryPolicy.retryDelay(from: httpResponse) {
```

Comment:
Maybe we can check for this `first` and if goes to `else` then call the super class (`RetryPolicy`) version? Then we don't have to handle and duplicate the computation for exponential backoff ourselves.

Reply:
Also, I think the gist of retrying is here and where this policy is installed right? The rest are mostly refactoring.